internal_types/src/entry.rs

```diff
@@ -1329,6 +1346,174 @@ impl TryFrom<Vec<u8>> for SequencedEntry {
+        match self.entry.as_ref() {
+            Some(e) => match e.operation_as_write().as_ref() {
+                Some(w) => w
+                    .partition_writes()
+                    .as_ref()
+                    .map(|w| w.iter().map(|fb| PartitionWrite { fb }).collect::<Vec<_>>()),
+                None => None,
+            },
+            None => None,
+        }
+    }
+
+    fn fb(&self) -> &entry_fb::SequencedEntry<'_> {
+        &self.fb
+    }
+
+    fn size(&self) -> usize {
+        self.fb._tab.buf.len()
+    }
+}
+
+#[self_referencing]
+#[derive(Debug)]
+pub struct Segment {
+    data: Vec<u8>,
+    #[borrows(data)]
+    #[covariant]
+    fb: entry_fb::Segment<'this>,
+    #[borrows(data)]
+    #[covariant]
+    sequenced_entries: Vec<BorrowedSequencedEntry<'this>>,
+}
+
+impl Segment {
+    pub fn new_from_entries(
+        segment_id: u64,
+        server_id: ServerId,
+        clock_value: Option<ClockValue>,
+        entries: &[Arc<OwnedSequencedEntry>],
+    ) -> Self {
+        let mut fbb = FlatBufferBuilder::new_with_capacity(1024);
+
+        let entries = entries
+            .iter()
+            .map(|e| {
+                let entry_bytes = fbb.create_vector_direct(
+                    e.fb()
+                        .entry_bytes()
+                        .expect("entry must be present in sequenced entry when initialized"),
+                );
+
+                entry_fb::SequencedEntry::create(
+                    &mut fbb,
+                    &entry_fb::SequencedEntryArgs {
+                        server_id: e.server_id().get_u32(),
+                        clock_value: e.clock_value().get_u64(),
+                        entry_bytes: Some(entry_bytes),
+                    },
+                )
+            })
+            .collect::<Vec<_>>();
+
+        let entries = fbb.create_vector(&entries);
+
+        let segment = entry_fb::Segment::create(
+            &mut fbb,
+            &entry_fb::SegmentArgs {
+                id: segment_id,
+                server_id: server_id.get_u32(),
+                consistency_high_water_clock: clock_value.map(|c| c.get_u64()).unwrap_or(0),
+                entries: Some(entries),
+            },
+        );
+
+        fbb.finish(segment, None);
+
+        let (mut data, idx) = fbb.collapse();
+
+        Self::try_from(data.split_off(idx))
+            .expect("Flatbuffer data for sequenced entry just constructed should be valid")
+    }
+
+    /// Returns the Flatbuffers struct for the Segment
+    pub fn fb(&self) -> &entry_fb::Segment<'_> {
+        self.borrow_fb()
+    }
+
+    /// Returns the serialized bytes for the Segment
+    pub fn data(&self) -> &[u8] {
+        self.borrow_data()
+    }
+
+    pub fn consistency_high_water_clock(&self) -> Option<ClockValue> {
+        match self.fb().consistency_high_water_clock().try_into() {
+            Ok(c) => Some(c),
+            _ => None,
+        }
```

Comment:
This can be a little shorter with:

```suggestion
        self.fb().consistency_high_water_clock().try_into().ok()
```